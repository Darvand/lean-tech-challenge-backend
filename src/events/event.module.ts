import { Module } from '@nestjs/common';
import { EventsController } from './presentation/event.controller';
import { EVENTS_PROVIDER } from './events.provider';
import { EventsMongoRepository } from './infraestructure/repositories/events-mongo.repository';
import { EventsService } from './application/events.service';
import { DatabaseModule } from 'src/database/database.module';
import { MongooseModule } from '@nestjs/mongoose';
import { Event, EventSchema } from './infraestructure/schemas/event.schema';

@Module({
  controllers: [EventsController],
  providers: [
    {
      provide: EVENTS_PROVIDER.EventRepository,
      useClass: EventsMongoRepository,
    },
    EventsService,
  ],
  imports: [
    DatabaseModule,
    MongooseModule.forFeature([{ name: Event.name, schema: EventSchema }]),
  ],
})
export class EventsModule {}
