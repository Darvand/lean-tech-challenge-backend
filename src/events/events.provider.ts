export const EVENTS_PROVIDER = {
  EventRepository: Symbol.for('EventRepository'),
};
