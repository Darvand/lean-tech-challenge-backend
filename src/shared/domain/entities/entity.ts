import { UUID } from '../value-objects/uuid.value-object';

export abstract class Entity<T> {
  protected readonly _id: UUID;
  public readonly props: T;

  constructor(props: T, id?: UUID) {
    this._id = id ? id : UUID.generate();
    this.props = props;
  }
}
