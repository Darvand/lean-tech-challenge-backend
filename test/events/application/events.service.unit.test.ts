import { Test, TestingModule } from '@nestjs/testing';
import { EventsService } from '../../../src/events/application/events.service';
import { EVENTS_PROVIDER } from '../../../src/events/events.provider';
import {
  EventStatus,
  EventStatusValues,
} from '../../../src/events/domain/value-objects/event-status.value-object';
import {
  TicketType,
  TicketTypesValues,
} from '../../../src/events/domain/value-objects/ticket-types.value-object';
import { EventAggregate } from '../../../src/events/domain/entities/event-aggregate.entity';
import { Title } from '../../../src/events/domain/value-objects/title.value-object';
import { DateValue } from '../../../src/shared/domain/value-objects/date-value.value-object';
import { Location } from '../../../src/events/domain/value-objects/location.value-object';
import { Number } from '../../../src/shared/domain/value-objects/number.value-object';
import { Benefits } from '../../../src/events/domain/value-objects/benefits.value-object';
import { EventTicket } from '../../../src/events/domain/entities/event-ticket.entity';
import { Description } from '../../../src/events/domain/value-objects/description.value-object';

describe('EventsService', () => {
  let service: EventsService;
  const findOneMock = jest.fn();

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      providers: [
        EventsService,
        {
          provide: EVENTS_PROVIDER.EventRepository,
          useValue: {
            findOne: findOneMock,
          },
        },
      ],
    }).compile();

    service = moduleFixture.get<EventsService>(EventsService);
  });
  describe('findOne', () => {
    const eventAggregate: EventAggregate = EventAggregate.create({
      userId: 'userId',
      title: Title.create('title'),
      description: Description.create('description'),
      status: EventStatus.create(EventStatusValues[0]),
      startDate: DateValue.createFromJSDate(new Date()),
      endDate: DateValue.createFromJSDate(new Date()),
      location: Location.create(0, 0),
      tickets: [
        EventTicket.create({
          type: TicketType.create(TicketTypesValues[0]),
          price: Number.create(10),
          available: Number.create(100),
          purchaseLimit: Number.create(10),
          benefits: Benefits.create(['benefit']),
        }),
      ],
    });
    it('should call findOne method from repository', async () => {
      findOneMock.mockResolvedValue(eventAggregate);
      const eventDto = await service.findOne(
        eventAggregate.id,
        eventAggregate.userId,
      );
      expect(findOneMock).toHaveBeenCalledTimes(1);
      expect(eventDto).toMatchSnapshot({
        id: expect.any(String),
        tickets: [{ id: expect.any(String) }],
      });
    });

    it('should throw ForbiddenException when user is not the owner', async () => {
      findOneMock.mockResolvedValue(eventAggregate);
      await expect(
        service.findOne(eventAggregate.id, 'otherUserId'),
      ).rejects.toThrow('You are not allowed to see this event');
    });
  });
});
