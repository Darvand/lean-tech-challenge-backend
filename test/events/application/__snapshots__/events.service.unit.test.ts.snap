// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventsService findOne should call findOne method from repository 1`] = `
{
  "description": "description",
  "endDate": "Feb 18",
  "id": Any<String>,
  "location": {
    "latitude": 0,
    "longitude": 0,
  },
  "startDate": "Feb 18",
  "status": "draft",
  "tickets": [
    {
      "available": 100,
      "benefits": [
        "benefit",
      ],
      "id": Any<String>,
      "price": 10,
      "purchaseLimit": 10,
      "type": "general-admission",
    },
  ],
  "title": "title",
}
`;
